### PR TITLE
DynamoDBToS3Operator - Add feature to export table to a point in time

### LINKED ISSUE #28830: Export DynamoDB table to S3 with PITR
### Description

Airflow provides the Amazon DynamoDB to Amazon S3 below.  
https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/operators/transfer/dynamodb_to_s3.html

Most of Data Engineer build their "export DDB data to s3" pipeline using "within the point in time recovery window".   
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodb.html#DynamoDB.Client.export_table_to_point_in_time

I appreciate if airflow has this function as a native function.  

### Use case/motivation

My daily batch job exports its data with pitr option. All of tasks is written by apache-airflow-providers-amazon except "export_table_to_point_in_time" task. 
"export_table_to_point_in_time" task only used the python operator. I expect I can unify the task as apache-airflow-providers-amazon library. 

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by utkarsharma2**
DynamoDBToS3Operator - Add a feature to export the table to a point in time.

closes: https://github.com/apache/airflow/issues/28830

### CODE REVIEW

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: @Taragolis What should be the value of `ClientToken`? 
Is it a random string to aid to idempotent nature of the method? or do I need to extract this value from the hook?

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (vincbeck) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]:
```suggestion
    :param export_time: Timestamp in the past from which to export table data (in seconds). The table export will be a snapshot of the table’s state at this point in time.
```

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (vincbeck) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: Instead of creating it here and passing to the private methods I would use `@cached_property` to create it. Here is an example: https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/operators/emr.py#L402

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (vincbeck) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: I would change the logic of this function. I would wait for the export to be done only if specified by the user with a new flag `wait_for_completion`. If `False`, then just trigger the export without waiting. If `True`, wait for the export but use a waiter in order to do so. Do not use a `while True`. See [here](https://github.com/apache/airflow/tree/main/airflow/providers/amazon/aws/waiters) how to create a custom waiter

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (vincbeck) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: In my opinion, you dont need to provide it

**Review by vincbeck — CHANGES_REQUESTED** [review]

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: Added a cached_property of the same.

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: @vincbeck want to understand it a bit, what would happen if we make the same call multiple times without passing `clientToken`? wouldn't it cause the same operation to occur multiple times?

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: Added this change

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: @vincbeck is there a better way to create/update the waiter config?

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: Thanks for the suggestions. :)
1. Updated code to implement `waiter` logic. This is much better now!
2. I feel if we add `wait_for_completion` it will be only applicable if a user is specifying `export_time` that is using the `ExportTableToPointInTime` API via a waiter. The older path cannot accommodate this logic since we are getting data to the worker node and pushing in a loop and would be a little inconsistent IMO. WDYT?

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (vincbeck) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: My understanding from the [doc](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodb/client/export_table_to_point_in_time.html) is if you make multiple calls with the same `clientToken`, the result will be the same (unless some corner cases, see doc for more details). If you dont provide it, you might get different result if the table content changed, which we want I think

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (vincbeck) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: The waiter config is better to be defined in a json file [here](https://github.com/apache/airflow/tree/main/airflow/providers/amazon/aws/waiters). You can take a look at this [PR](https://github.com/apache/airflow/pull/29761) as example on how to create and use these custom waiters. I hope it helps!

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (vincbeck) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: That makes sense :) I agree then, introducing a flag `wait_for_completion` here does not make sense

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: Thanks for pointing me in the right direction, much helpful :). Made changes accordingly.

**Comment by utkarsharma2** [comment]
@eladkal @o-nikolas Please review

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (vincbeck) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py:119]: Please add it to docstring

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (vincbeck) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: Unused?

**Review by vincbeck — CHANGES_REQUESTED** [review]

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py:119]: Added.

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: Removed.

**Review by vincbeck — APPROVED** [review]

**Comment by utkarsharma2** [comment]
FYI - Test cases are failing because of the following issue - https://github.com/apache/airflow/issues/30613

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: Ya, that makes sense. :)

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (ephraimbuddy) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py:129]: What will happen if the export_time is a time in the future? Might be worth protecting against this

Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py:129]: @ephraimbuddy Good point, I can add a check for that.

Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py:129]: @ephraimbuddy I have added a check for future DateTime. Please check.

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (o-nikolas) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: Just for my own education: Usually you see a waiter used after an initial api action. For example an api request to create some resource, and and then the waiter function will call an api like describe periodically in a loop until the desired state is returned (CREATED maybe in my example). But here you're just using the waiter to start the export and to wait until it's completed, very neat! But I'm curious about idempotency, currently the waiter is configured to poll every 30s, so if the export takes longer than 30s, a second request to the api will be sent, is that okay? Are there idempotency or consistency issues with this? Or have I entirely misunderstood this change? :laughing:

**Comment by potiuk** [comment]
Needs conflict resolving too.

File: airflow/providers/amazon/aws/waiters/dynamodb.json
Comment (ferruzzi) [inline airflow/providers/amazon/aws/waiters/dynamodb.json:12]: According to my notes, I think the current issue is because these "argument" fields need to be all lowercase.  Documentation on the BOTO side for these is pretty hard to find, I might be mistaken, but if you are still running into that ` 'dynamodb.ServiceResource' object has no attribute 'describe_export'` issue. 

 If that is the case, can you please also update the waiters README to make that more obvious??

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (o-nikolas) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: For posterity: Utkarsh has updated the code to fix this and use a more typical API request + waiter approach

File: airflow/providers/amazon/aws/waiters/dynamodb.json
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/waiters/dynamodb.json:12]: @ferruzzi I'm still facing the same issue post changes to Arguments. Updated json config looks like:
```
{
    "version": 2,
    "waiters": {
        "export_table": {
            "operation": "DescribeExport",
            "delay": 30,
            "maxAttempts": 60,
            "acceptors": [
                {
                    "matcher": "path",
                    "expected": "COMPLETED",
                    "argument": "exportdescription.exportstatus",
                    "state": "success"
                },
                {
                    "matcher": "path",
                    "expected": "FAILED",
                    "argument": "exportdescription.exportstatus",
                    "state": "failure"
                },
                {
                    "matcher": "path",
                    "expected": "IN_PROGRESS",
                    "argument": "exportdescription.exportstatus",
                    "state": "retry"
                }
            ]
        }
    }
}
```
Would it make sense to change the logic to polling on the `DescribeExport` API, until there is a fix for the waiters in botocore?

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: @o-nikolas Thanks for the update :)

File: airflow/providers/amazon/aws/waiters/dynamodb.json
Comment (ferruzzi) [reply airflow/providers/amazon/aws/waiters/dynamodb.json:12]: I was mistaken on the lowercase suggestion.  Looking through a handful of other examples of official waiters in the botocore package, (see [EC2](https://github.com/boto/botocore/blob/develop/botocore/data/ec2/2016-11-15/waiters-2.json#L81) for example)  they all seem to match the case exactly as the [boto API docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodb/client/describe_export.html) have it,, so in this instance PascalCase would be the right answer.  Sorry about that.

I have one last idea if you are willing to try it out, but I won't block the PR longer for this.

I think the root issue is that the DynamoDB hook is creating a `resource_type` instead of a `client_type` and, as the erro message says, the ServiceResource doesn't have a method called `describe_export`.  But the BaseAwsHook `get_waiter` method accepts a client as a parameter, so we could try that.  Something like:

```
waiter = self.hook.get_waiter(waiter_name=CUSTOM_WAITER_NAME, client=boto3.client('dynamodb'))
waiter.wait(ExportArn=response.get("ExportArn"))
```


If that doesn't work, then I'm willing to concede this one and we can do a polling loop here until I et time to look deeper into why this one won't cooperate.

Comment (utkarsharma2) [reply airflow/providers/amazon/aws/waiters/dynamodb.json:12]: @ferruzzi This worked! Thanks :)

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]:
```suggestion
    :param export_time: Time in the past from which to export table data, counted in seconds from the start of
```

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py:151]: Not for now, but boy is this ugly, we should make a better answer for fetching the client at some point.

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: (Non-blocking) I appreciate the constant here instead of a magic string, but I think in this case I'd likely inline it.  I feel like we usually use waiter names directly and don't have a problem with it here.  But I'll leave it up to you.

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: (Non-blocking)  The pre-commit/static checks broke this into three lines because of the trailing comma.  If you remove that, this will get collapsed into a one-liner.

File: airflow/providers/amazon/aws/waiters/README.md
Comment (ferruzzi) [inline airflow/providers/amazon/aws/waiters/README.md:102]: Thanks for adding this.  As we discussed on Slack, we could tackle this a few ways, but none of them should be considered blocking for this PR.  As a temporary Airflow-based fix, I like the idea of adding come logic to Base AWS Hook's "get waiter" method that checks if the cal is being made against a Resource and creates a client connection to handle it if that is the case.  Longer term, it would be nice to seer why botocore doesn't support this natively and maybe submit a PR to add it... but the hook solution is a something we could do quickly in code we own.

**Review by ferruzzi — APPROVED** [review]
Left a couple of comments and non-blocking suggestion, but I like how this turned out.  The PascalCase suggestion needs to be fixed, the other two are just thoughts you can feel free to revise or ignore.

 I really appreciate your patience with the waiter shenanigans, we both learned something on that!

**Review by ferruzzi — APPROVED** [review]
Nice :+1:

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: Ya, looks like overkill, since we are just using it at a single place.

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: Done.

File: airflow/providers/amazon/aws/waiters/README.md
Comment (utkarsharma2) [reply airflow/providers/amazon/aws/waiters/README.md:102]: Yup, I'm planning to address Airflow fix in a separate PR :)

Comment (utkarsharma2) [reply airflow/providers/amazon/aws/waiters/README.md:102]: @ferruzzi @o-nikolas @vincbeck Please check PR - https://github.com/apache/airflow/pull/30897

File: airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py]: :heart:

**Comment by ferruzzi** [comment]
If the CI is happy, I'm happy :+1:   Thanks for sticking with this one.

**Comment by potiuk** [comment]
CI is not happy. @utkarsharma2 -> you need to fix those .

**Review by potiuk — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Comment by ferruzzi** [comment]
@utkarsharma2 Did you actually try this live?   After the merge, the system test fails because [this addition](https://github.com/apache/airflow/pull/30501/files#diff-849549a5587d7b4e2e386cb23d5b8da9cddfd5b88b80fb931c36cb931e943e6fR131-R139) is missing the file_size paramter, and after adding that in it fails with the message

```
ERROR    airflow.task:taskinstance.py:1900 Task failed with exception
Traceback (most recent call last):
  File "/opt/airflow/airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py", line 142, in execute
    self._export_table_to_point_in_time()
  File "/opt/airflow/airflow/providers/amazon/aws/transfers/dynamodb_to_s3.py", line 157, in _export_table_to_point_in_time
    ExportFormat=self.export_format,
  File "/usr/local/lib/python3.7/site-packages/botocore/client.py", line 530, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python3.7/site-packages/botocore/client.py", line 960, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (ValidationException) when calling the ExportTableToPointInTime operation: One or more parameter values were invalid: tableArn is not a valid ARN
```Here

[Here](https://github.com/apache/airflow/pull/30501/files#diff-5a4b3715577fad0a7bbf023f861a5f4f1e0fb1abb4b2a71ecb804a068d9de4edR153) you are passing`dynamodb_table_name` instead of the ARN.   I'm not sure how this would have worked, and I probably should have noticed that in the review.

**Comment by utkarsharma2** [comment]
@ferruzzi Yes I did, I was passing table ARN in place of the table name.
```
  dynamodb_to_s3_operator = DynamoDBToS3Operator(
      task_id="dynamodb_to_s3",
      dynamodb_table_name="arn:aws:dynamodb:us-east-1:633294268925:table/test",
      s3_bucket_name="tmp9",
      file_size=4000,
      export_time=datetime.now(),
      aws_conn_id="aws_default",
      s3_key_prefix="test"
  )
```
Maybe we can have different param for getting Table ARN?

**Comment by utkarsharma2** [comment]
@ferruzzi @o-nikolas Raised a new PR with the Fix Please review - https://github.com/apache/airflow/pull/31142

**Comment by ferruzzi** [comment]
Cool, thanks!  I'll take a look later today when I get a moment :+1:

**Pull request merged**
